### PR TITLE
[Lang] Fix integral type promotion rules (e.g., u8 + u8 now leads to u8 instead of i32)

### LINKED ISSUE #4613: Why are there precision warnings when no type changes (target = u8, value = u8)?
Hi, I jsut tried doing bit operations and got some losing precision warnings that I could not understand since the data type wasn't changed. Is this a bug?
Besides, why there was another warning (`Atomic bit_or (i32 to u8)`) when I had explicitly cast the type before the bit operation in `set_bit()`?
```
@ti.func
def set_bit(switches, i, dt):
    switches  |= (ti.cast(1, dt) << i)
    return switches

@ti.kernel
def test():
    for _ in range(1):
        for i in range(8):
            eight_flags = ti.cast(0, ti.u8)
            eight_flags = set_bit(eight_flags, i, ti.u8)
            print(ti.cast(eight_flags, ti.i32))
test()
```

```
[Taichi] version 0.9.1, llvm 10.0.0, commit e2e0e669, linux, python 3.7.4
[Taichi] Starting on arch=x64
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2592] Local store may lose precision (target = u8, value = u8), at
2592
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2595] Local store may lose precision (target = u8, value = u8), at
2595
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@67] [$2603] Atomic bit_or (i32 to u8) may lose precision, at

[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2604] Local store may lose precision (target = u8, value = u8), at
2604
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2606] Local store may lose precision (target = u8, value = u8), at
2606
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2592] Local store may lose precision (target = u8, value = u8), at
2592
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2595] Local store may lose precision (target = u8, value = u8), at
2595
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2604] Local store may lose precision (target = u8, value = u8), at
2604
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2606] Local store may lose precision (target = u8, value = u8), at
2606
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2592] Local store may lose precision (target = u8, value = u8), at
2592
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2595] Local store may lose precision (target = u8, value = u8), at
2595
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2604] Local store may lose precision (target = u8, value = u8), at
2604
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2606] Local store may lose precision (target = u8, value = u8), at
2606
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2592] Local store may lose precision (target = u8, value = u8), at
2592
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2595] Local store may lose precision (target = u8, value = u8), at
2595
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2604] Local store may lose precision (target = u8, value = u8), at
2604
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2606] Local store may lose precision (target = u8, value = u8), at
2606
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$22] Local store may lose precision (target = u8, value = u8), at
22
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$2621] Local store may lose precision (target = u8, value = u8), at
2621
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at
27
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at
27
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at
27
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.275 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at
27
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at
27
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at
27
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at
27
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$19] Local store may lose precision (target = u8, value = u8), at
19
[W 03/24/22 13:19:51.276 148735] [type_check.cpp:visit@142] [$27] Local store may lose precision (target = u8, value = u8), at 27
```

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = closes #4613
Related issue = #2196

It's interesting that in C++ `i8 + i8 = i32`.

**Type promotion table:**
Format: a + b = old_type -> new_type. Only some are changed.

```
f32 + f32 = f32
f32 + f64 = f64
f32 + i8 = f32
f32 + i16 = f32
f32 + i32 = f32
f32 + i64 = f32
f32 + u8 = f32
f32 + u16 = f32
f32 + u32 = f32
f32 + u64 = f32
f64 + f32 = f64
f64 + f64 = f64
f64 + i8 = f64
f64 + i16 = f64
f64 + i32 = f64
f64 + i64 = f64
f64 + u8 = f64
f64 + u16 = f64
f64 + u32 = f64
f64 + u64 = f64
i8 + f32 = f32
i8 + f64 = f64
i8 + i8 = i32 -> i8
i8 + i16 = i32 -> i16
i8 + i32 = i32
i8 + i64 = i64
i8 + u8 = i32 -> u8
i8 + u16 = i32 -> u16
i8 + u32 = u32
i8 + u64 = u64
i16 + f32 = f32
i16 + f64 = f64
i16 + i8 = i32 -> i16
i16 + i16 = i32 -> i16
i16 + i32 = i32
i16 + i64 = i64
i16 + u8 = i32 -> i16
i16 + u16 = i32 -> u16
i16 + u32 = u32
i16 + u64 = u64
i32 + f32 = f32
i32 + f64 = f64
i32 + i8 = i32
i32 + i16 = i32
i32 + i32 = i32
i32 + i64 = i64
i32 + u8 = i32
i32 + u16 = i32
i32 + u32 = u32
i32 + u64 = u64
i64 + f32 = f32
i64 + f64 = f64
i64 + i8 = i64
i64 + i16 = i64
i64 + i32 = i64
i64 + i64 = i64
i64 + u8 = i64
i64 + u16 = i64
i64 + u32 = i64
i64 + u64 = u64
u8 + f32 = f32
u8 + f64 = f64
u8 + i8 = i32 -> u8
u8 + i16 = i32 -> i16
u8 + i32 = i32
u8 + i64 = i64
u8 + u8 = i32 -> u8
u8 + u16 = i32 -> u16
u8 + u32 = u32
u8 + u64 = u64
u16 + f32 = f32
u16 + f64 = f64
u16 + i8 = i32 -> u16
u16 + i16 = i32 -> u16
u16 + i32 = i32
u16 + i64 = i64
u16 + u8 = i32 -> u16
u16 + u16 = i32 -> u16
u16 + u32 = u32
u16 + u64 = u64
u32 + f32 = f32
u32 + f64 = f64
u32 + i8 = u32
u32 + i16 = u32
u32 + i32 = u32
u32 + i64 = i64
u32 + u8 = u32
u32 + u16 = u32
u32 + u32 = u32
u32 + u64 = u64
u64 + f32 = f32
u64 + f64 = f64
u64 + i8 = u64
u64 + i16 = u64
u64 + i32 = u64
u64 + i64 = u64
u64 + u8 = u64
u64 + u16 = u64
u64 + u32 = u64
u64 + u64 = u64
```

### CODE REVIEW

File: taichi/ir/type_factory.cpp [hunk 1]
```diff
@@ -103,6 +103,20 @@ PrimitiveType *TypeFactory::get_primitive_int_type(int bits, bool is_signed) {
...
+    real_type = get_primitive_type(PrimitiveTypeID::f16);
+  } else if (bits == 32) {
+    real_type = get_primitive_type(PrimitiveTypeID::f32);
+  } else if (bits == 64) {
+    real_type = get_primitive_type(PrimitiveTypeID::f64);
+  } else {
+    TI_ERROR("No primitive real type has {} bits", bits);
+  }
+  return real_type->cast<PrimitiveType>();
+}
```
Comment (yuanming-hu) [inline taichi/ir/type_factory.cpp:118]: Not really useful in this PR. Maybe we can save it for the future, or I can remove it.

Comment (strongoier) [reply taichi/ir/type_factory.cpp:118]: Saving it here is fine.

**Review by strongoier — APPROVED** [review]
LGTM!!

I was previously stuck at this issue because I thought we must follow C in arithmetic operations. However this is not a correct assumption...

BTW I think we need to update the doc as well: https://github.com/taichi-dev/taichi/blob/022456def2be5f592488ae983580aebb2f61525b/docs/lang/articles/basic/type.md?plain=1#L103

**Comment by yuanming-hu** [comment]
Thank you @strongoier! 

>  I thought we must follow C in arithmetic operations. However this is not a necessary assumption...

I was also stuck a bit here, until I realized that LLVM uses a similar rule as ours:

```C++
char add(char a, char b) {
    return a + b;
}
```

```llvm
define signext i8 @add(i8 signext %0, i8 signext %1) local_unnamed_addr #0 {
  %3 = add i8 %1, %0
  ret i8 %3
}
```

Could you help create an issue for the doc update?

One more thing I would like to point out:

Currently, all binary operators follow the current promotion rule. But there are exceptions. For example, `i8 << u32` may end up with `i8`, but currently it ends up with `u32`. We need to reach a consensus here: should we simply take the left-hand side type as the shift operator outputs?

**Comment by strongoier** [comment]
> Could you help create an issue for the doc update?

#4793.

> We need to reach a consensus here: should we simply take the left-hand side type as the shift operator outputs?

I agree. #4795.

**Pull request merged**